File: backend/services/agent-runner/src/stigmer_runner/worker/activities/graphton/environment.py
"""Environment variable resolution for Graphton agent execution.

Resolves the merged environment from ExecutionContext — the single source
of truth for execution-scoped variables.  The workflow injects all
resolved variables into ExecutionContext before the activity starts, and
it persists until execution termination.
"""

from __future__ import annotations

import dataclasses
import logging
from typing import Any


@dataclasses.dataclass(frozen=True)
class EnvironmentResult:
    """Output of :func:`resolve_environment`."""

    merged_env_vars: dict[str, str]
    secret_keys: set[str]


async def resolve_environment(
    *,
    execution_context_client: Any,
    execution_id: str,
    logger: logging.Logger,
) -> EnvironmentResult:
    """Resolve merged environment variables from ExecutionContext.

    ExecutionContext is created by the workflow with pre-merged env vars
    before the activity starts.  If no ExecutionContext exists (e.g. the
    agent has no secrets or environment variables configured), an empty
    result is returned so the execution can proceed without env setup.

    Args:
        execution_context_client: ``ExecutionContextClient`` instance.
        execution_id: Current execution ID.
        logger: Activity logger.

    Returns:
        ``EnvironmentResult`` with merged vars and secret key set.
    """
    exec_ctx = await execution_context_client.try_get_by_execution_id(
        execution_id,
    )

    if not exec_ctx:
        logger.info(
            "No ExecutionContext found for execution %s — "
            "proceeding with empty environment.",
            execution_id,
        )
        return EnvironmentResult(merged_env_vars={}, secret_keys=set())

    merged_env_vars: dict[str, str] = {}
    secret_keys: set[str] = set()

    if exec_ctx.spec.data:
        for key, exec_value in exec_ctx.spec.data.items():
            merged_env_vars[key] = exec_value.value
            if exec_value.is_secret:
                secret_keys.add(key)

    logger.info(
        "Resolved environment from ExecutionContext: "
        "context_id=%s, env_count=%d, secret_count=%d",
        exec_ctx.metadata.id,
        len(merged_env_vars),
        len(secret_keys),
    )

    return EnvironmentResult(
        merged_env_vars=merged_env_vars,
        secret_keys=secret_keys,
    )

File: backend/services/runner/src/activities/execute-deep-agent/environment.ts
/**
 * Environment variable resolution for deep agent execution.
 *
 * Resolves the merged environment from ExecutionContext — the single
 * source of truth for execution-scoped variables. The workflow injects
 * all resolved variables into ExecutionContext before the activity starts.
 *
 * The returned result separates plaintext keys from secret keys so that
 * downstream consumers (e.g., StatusBuilder in Phase 3b) can redact
 * secret values in user-facing displays.
 */

import type { StigmerClient } from "../../client/stigmer-client.js";

export interface EnvironmentResult {
  readonly mergedEnvVars: Record<string, string>;
  readonly secretKeys: ReadonlySet<string>;
}

/**
 * Fetch the ExecutionContext for the given execution and extract the
 * merged environment variables.
 *
 * If no ExecutionContext exists (e.g., the agent has no environment or
 * secrets configured), returns an empty result so execution can proceed.
 */
export async function resolveEnvironment(
  client: StigmerClient,
  executionId: string,
): Promise<EnvironmentResult> {
  // A desktop runner exchanges its bootstrap credential for a token scoped to
  // this execution's session, so cloud's decrypt gate binds the read (#156).
  // No-op for cloud sandbox and OSS runners. A failed exchange throws and
  // fails the activity: the bootstrap credential no longer decrypts
  // (stigmer-cloud#218), so proceeding would resolve redacted placeholders.
  const scopedToken = await client.acquireScopedRunnerToken({
    agentExecutionId: executionId,
  });

  let execCtx;
  try {
    execCtx = await client.getExecutionContextByExecutionId(executionId, scopedToken);
  } catch (err: unknown) {
    // ConnectError uses numeric Code.NotFound (5); match the pattern
    // from execute-cursor/env-resolver.ts.
    const code = (err as { code?: number | string })?.code;
    if (code === 5 || code === "not_found" || code === "NOT_FOUND") {
      console.log(
        `[env] No ExecutionContext found for execution ${executionId} — ` +
        `proceeding with empty environment.`,
      );
      return { mergedEnvVars: {}, secretKeys: new Set() };
    }
    throw err;
  }

  const mergedEnvVars: Record<string, string> = {};
  const secretKeys = new Set<string>();

  const data = execCtx.spec?.data;
  if (data) {
    for (const [key, execValue] of Object.entries(data)) {
      mergedEnvVars[key] = execValue.value;
      if (execValue.isSecret) {
        secretKeys.add(key);
      }
    }
  }

  console.log(
    `[env] Resolved environment: env_count=${Object.keys(mergedEnvVars).length}, ` +
    `secret_count=${secretKeys.size}`,
  );

  return { mergedEnvVars, secretKeys };
}
